File: LeetCode-py/404_sumOfLeftLeaves.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.root = None

    def createTree(self, nodeList):
        if nodeList[0] == None:
            return None
        root = TreeNode(nodeList[0])
        i = 1
        Nodes = [root]
        for node in Nodes:
            if node != None:
                node.left = TreeNode(nodeList[i]) if nodeList[i] != None else None
                Nodes.append(node.left)
                i += 1
                if i == len(nodeList): return root
                node.right = TreeNode(nodeList[i]) if nodeList[i] != None else None
                Nodes.append((node.right))
                i += 1
                if i == len(nodeList): return root

    def preorderTraversal(self, root: TreeNode):
        if root!=None:
            print(root.val, end=' ')
            self.preorderTraversal(root.left)
            self.preorderTraversal(root.right) 

class Solution:
    @classmethod
    def sumOfLeftLeaves(self, root: TreeNode):
        self.summ = 0
        def dfs(root, left):
            if not root:
                return
            if(not root.left and not root.right and left):
                self.summ += root.val
            dfs(root.left, True)
            dfs(root.right, False)
        dfs(root, False)
        return self.summ


if __name__ == "__main__":
    t = Tree()
    t.root = t.createTree([3,9,20,None,None,15,7])
    print("先序遍历：")
    t.preorderTraversal(t.root)
    print("\n左叶子之和：", Solution.sumOfLeftLeaves(t.root))
    

File: LeetCode-c++/404_sumOfLeftLeaves.cpp
#include<vector>
#include<queue>
#include<iostream>
#include<string>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode * left;
    TreeNode * right;
    TreeNode(int x): val(x), left(NULL), right(NULL) {}
};

class Tree
{
    public:
    TreeNode * root;
    Tree(vector<int>& nodeList)
    {
        root = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        createTree(nodeList);
    }

    TreeNode* createTree(vector<int> nodeList)
    {
        int len = nodeList.size();
        if(len==0 || !nodeList[0])
            return nullptr;
        root = new TreeNode(nodeList[0]);
        queue<TreeNode*> que;
        que.push(root);
        int i=1;
        while (i<len)
        {
            TreeNode* node = que.front();
            que.pop();
            if(nodeList[i])
            {
                TreeNode* n = new TreeNode(nodeList[i]);
                node->left = n;
                que.push(n);
            }
            if(++i==len) break;
            if(nodeList[i])
            {
                TreeNode* n = new TreeNode(nodeList[i]);
                node->right = n;
                que.push(n);
            }
            if(++i==len) break;
        }
    }

    void Traversal(int order = 0)
    {
        switch (order)
        {
        case 0:
            cout<< "先序遍历：";
            preorderTraversal(root);
            break;
        case 1:
            cout<< "中序遍历：";
            inorderTraversal(root);
            break;
        case 2:
            cout<< "后序遍历：";
            postorderTraversal(root);
            break;
        default:
            cout << "遍历选项不存在"<< endl;
            break;
        }
        cout << endl;
    }
    void preorderTraversal(TreeNode * root)
    {
        if(!root)
            return;
        cout<< root->val <<' ';
        preorderTraversal(root->left);
        preorderTraversal(root->right);
    }
    void inorderTraversal(TreeNode * root)
    {
        if(!root)
            return;
        preorderTraversal(root->left);
        cout<< root->val <<' ';
        preorderTraversal(root->right);
    }
    void postorderTraversal(TreeNode * root)
    {
        if(!root)
            return;
        preorderTraversal(root->left);
        preorderTraversal(root->right);
        cout<< root->val <<' ';
    }
    // private:
    
};

class Solution
{
    public:
    int sumOfLeft = 0;
    int sumOfLeftLeaves(TreeNode* root)
    {
        dfs(root, false);
        return sumOfLeft;
    }
    void dfs(TreeNode* root, bool left)
    {
        if(!root) return;
        if(!root->left && !root->right && left)
            sumOfLeft += root->val;
        dfs(root->left, true);
        dfs(root->right, false);
    }
};

int main()
{
    vector<int> nodeList = {3, 9, 20, NULL, NULL, 15, 7};
    Tree t(nodeList);
    t.Traversal();
    Solution solution = Solution();
    cout << "左叶子之和：" << solution.sumOfLeftLeaves(t.root);
    system("pause");
    return 0;
}